### PR TITLE
fix: update benchmarks

### PR DESCRIPTION
**Pull request by ntn-x2**
Part of https://github.com/KILTprotocol/ticket/issues/3650, built on top of https://github.com/KILTprotocol/kilt-node/pull/787.

What started as a simple fix for the web3name pallet benchmarks, turned out to be a bigger changeset due to a couple of bugs found in the Substrate code, which are mentioned in the self-review I gave myself [below](https://github.com/KILTprotocol/kilt-node/pull/790#pullrequestreview-2427330495).

In a gist, what I did is:

1. Update the benchmarking logic for the web3name pallet to delegate the name creation to a `BenchmarkHelper` type.
2. Implement the trait above for both web3names (which uses the old implementation) and the dotnames
3. Work around the Substrate bug for our pallets as well as the `pallet_collective` and `pallet_membership` pallets which we deploy multiple times
4. Add the benchmarking of the `pallet_membership` deployments of the `TipsMembership` pallet

### CODE REVIEW

**Comment by Ad96el** [comment]
I looked into the benchmark error Pr on the polkadot-sdk, but I am struggling to understand what the underlying issue is. For pallets with multiple instances in the runtime, the benchmarking was executed for each instance with all different config configuration?

File: runtimes/peregrine/src/weights/pallet_dot_names.rs [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+// --output=./runtimes/peregrine/src/weights/
+
+#![cfg_attr(rustfmt, rustfmt_skip)]
+#![allow(unused_imports)]
+#![allow(clippy::as_conversions)]
+
+use frame_support::{traits::Get, weights::Weight};
+use sp_std::marker::PhantomData;
+
+// TODO: Remove once the benchmarking stuff fully supports pallet instances.
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply runtimes/peregrine/src/weights/pallet_dot_names.rs:49]: Do you mean the renaming of the import has to be done manually? If yes, shouldn't this also be done for other pallets with multiple instances, such as the collective pallet?

**Review by Ad96el — COMMENTED** [review]
two minior questions

File: runtimes/peregrine/src/lib.rs [hunk 2]
```diff
@@ -564,7 +565,26 @@ impl pallet_membership::Config<TechnicalMembershipProvider> for Runtime {
...
 	type MembershipChanged = TechnicalCommittee;
 	type MaxMembers = constants::governance::TechnicalMaxMembers;
-	type WeightInfo = weights::pallet_membership::WeightInfo<Runtime>;
+	type WeightInfo = weights::pallet_technical_membership::WeightInfo<Runtime>;
+}
+
+// Implementation of `MembershipChanged` equivalent to using `()` but that
+// returns `Some(AccountId::new([0; 32]))` in `get_prime()` only when
+// benchmarking. TODO: Remove once we upgrade with a version containing the fix: https://github.com/paritytech/polkadot-sdk/pull/6439
+pub struct MockMembershipChangedForBenchmarks;
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply runtimes/peregrine/src/lib.rs:574]: No. The fix was written few days ago when we found the issue, so I expect it to be released only from the next polkadot version and the next ones.

File: runtimes/peregrine/src/weights/pallet_dot_names.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ntn-x2) [reply runtimes/peregrine/src/weights/pallet_dot_names.rs:49]: We do that, don't we?

**Comment by ntn-x2** [comment]
> I looked into the benchmark error Pr on the polkadot-sdk, but I am struggling to understand what the underlying issue is. For pallets with multiple instances in the runtime, the benchmarking was executed for each instance with all different config configuration?

That is exactly the issue. Plus, a related error and a somewhat consequence of this is that the calculated weights are actually the interpolation of the results combined, for each instance, that is why we had to rename the pallet of each instance, so that one instance run would not interfere with the others. This is a pretty serious bug in Substrate, and it should have been caught way way earlier.

File: runtimes/peregrine/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ad96el) [reply runtimes/peregrine/src/lib.rs:574]: Alright I see

File: runtimes/peregrine/src/weights/pallet_dot_names.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ad96el) [reply runtimes/peregrine/src/weights/pallet_dot_names.rs:49]: Oh, yes you are right

**Comment by Ad96el** [comment]
> That is exactly the issue. Plus, a related error and a somewhat consequence of this is that the calculated weights are actually the interpolation of the results combined, for each instance, that is why we had to rename the pallet of each instance, so that one instance run would not interfere with the others. This is a pretty serious bug in Substrate, and it should have been caught way way earlier.


I see thanks for the conformation

**Review by Ad96el — APPROVED** [review]

**Pull request merged**